### ISSUE TITLE
안드로이드기기(LTE)에서 외부 iotivity server에접근하고자합니다

### ISSUE THREAD

**Issue by sangilIm** [Einstrasse/graduation_work#2]
Simple client 예제를 활용해서 가능한지 여쭈어봐되될까요? 아니면 iotivity  cloud?란 것을 반드시 활용해야하나요?
현재 로컬네트워크에서 simpleClient , Server  테스트는 완료한상태입니다

**Comment by Einstrasse** [comment]
네, IoTivity 프레임워크에서는 IoTivity Cloud를 이용해서 외부 네트워크로부터 접근하도록 설계되어 있습니다.
아니면 프록시 역할을 하는 노드를 로컬에 두고, 해당 노드를 통해 간접적으로 접근하도록 직접 만드는 방법도 있는데, 해당 기능이 IoTivity Cloud가 하는 역할과 중복되므로, IoTivity Cloud를 사용하는것을 추천드립니다.

**Comment by sangilIm** [comment]
답변감사드립니다 ^^

그러면 Simple Client & Simple Server의 경우 같은 Local Network 안에서 Server에서 Resource를 자체적으로 등록하고 Client가 Resource를 발견하는 방식이었지만

Iotivity Cloud를 이용할 경우 기존 Server역할을 하는 장치가 Iotivity Cloud 서버에 resource를 등록하고 Client 역할을 하는 장치가 resource를 iotivity Cloud 서버에 요청하면 이를 Iotivity Cloud가 중계해주는 역할을 하는 것인가요?

그리고 Simple Clinet & Server 환경에서 Dest IP를 입력하는 란이 없어서 찾아보니 통신을 Multicast 방식으로 한다고하던데 API 내부적으로 고정된 Multicast 주소를 사용하나요??

마지막으로 죄송하지만 Cloud 테스트할 때 궁금한점이 있으면 질문드려도 될까요?? 
현재 Cloud Server 는 ubuntu에서 실행하였는데 Client로 Server에 접근하는 방법을 잘 모르겠어서 더 테스트 진행해보고 있습니다 ㅜㅜ 정말 모르겠으면 질문드릴게요

**Comment by Einstrasse** [comment]
네 말씀하신것처럼 iotivity cloud는 로컬에 설치된 리소스의 프록시역할을 합니다. 그 기능이 iotivity에서 제공하는 feature중에 resource directory기능이 해당 프록시역할 비스무리하게 하는 것인데, iotivity cloud는 해당 feature를 포함한 복합적인 feature입니다.

로컬환경에서 아이피주소 없이 리소스를 찾을 수 있는 것은, 내부적으로 멀티캐스트를 활용하며 이는 iotivity의 feature중 하나로 보시면됩니다. 가용한 transport 종류(ble bluetooth ipv4 ipv6등)을 지정하고, 리소스를 식별할수있는 정보(텍스트 형태의 태그나, url기반이었던것 같은데 정확히는 기억나지않습니다)만으로 정확한 network address를 모르더라도 리소스를 discovery할수있습니다. 멀티캐스트 주소 대역은 모르더라도 IoT 환경 구축에는 상관없으며 (몰라도 된다는 것 자체가 iotivity feature입니다)때문에, 자세하게 파보지는않았습니다.

또 궁금하신 것이 있으시면 제가 아는범위안에서는 답변드리겠습니다

https://github.com/Einstrasse/graduation_work/blob/master/cloud/readme.md
제가 정리해놓은것과, iotivity wiki페이지, iotivity cloud 샘플을 활용해보시면 이해에 도움될 것 같습니다.

**Comment by sangilIm** [comment]
답변 감사드립니다
혹시 transport를 ble 방식으로 테스트 해보셨나요? 현재 2대의 안드로이드기기로 한쪽에서는 peripheral 한쪽은 centeral 로 iotivity  ble 를 활용하여.통신테스트 해보려고하는데 구글에도 예제가없고 CaInterface 클래스에있는 starLeAdvertising 메소드를 호출해도 LE스캐너에 찍히는것도없네요 혹시 아시나요?

**Comment by Einstrasse** [comment]
ble는 안써봐서 잘 모르겠네요

**Comment by sangilIm** [comment]
안녕하세요 

아래 명령어로 Client Build를 진행하였는데요 

~iotivity $ scons cloud WITH_TCP=yes RELEASE=yes TARGET_TRANSPORT=IP WITH_CLOUD=yes WITH_MQ=PUB,SUB

가이드에서 보면 아래 경로에
~iotivity/out/linux/{ARCH}/release/cloud/samples/client 

아래 샘플 파일들이 있다고하는데
You can find aircon_controlee, aircon_controller, group_invite_sample, mq_publisher, mq_subscriber, thin_room_light. 

저는 왜 cloud 폴더가 없을 까요?? iotivity 1.3.0 버전으로 빌드하였습니다.
root@sangil:/home/sangil/iotivity-1.3.0-android/out/linux/i386/release# ls
libc_common.a                   libmbedtls.a       liboc.so
libcoap.a                       libmbedx509.a      libocsrm.a
libconnectivity_abstraction.a   liboc_logger.so    liboctbstack.a
libconnectivity_abstraction.so  libocpmapi.a       liboctbstack.so
liblogger.a                     libocpmapi.so      libroutingmanager.a
libmbedcrypto.a                 libocprovision.so  resource

뭔가 있을까 싶어 resource 폴더를 뒤져보았는데

cloudClient 실행파일이 있더라구요 

root@sangil:/home/sangil/iotivity-1.3.0-android/out/linux/i386/release/resource/provisioning/examples# ls
cloudAuth.o  cloudClient  cloudClient.o  cloud.dat  cloudWrapper.o  rootca.crt




그래서 실행 시켜보니
root@kbell:/home/kbell/iotivity-1.3.0-android/out/linux/i386/release/resource/provisioning/examples# ./cloudClient 

아래와 같이 옵션들과 원하는 메뉴를 선택 하는 창이 등장하였는데
문제는 어떤 절차로 서버와 연결하고 통신해야하는지 감히 잡히지 않습니다..

Connecting to: 127.0.0.1:5683
via auth provider: github
srv file: ./cloud.dat
CoAP prefix: coap+tcp://
-----------------------------------------------------------
************************************************************
****************** Cloud Client Requests *******************
************************************************************
** AUTHORIZATION
** 1 - Sign Up request
** 2 - Sign In request
** 3 - Sign Out request
** SETTINGS 
** 4 - Change default host
** 5 - Change default port
** 7 - Change default auth provider
** 8 - Change TLS cipher suite to RSA
** 9 - Save Trust Cert. Chain into Cred of SVR
** 10 - Change Protocol type (CoAP/CoAPs)
** CERTIFICATE REQUEST
** 19 - Certificate Request
** CRL
** 20 - CRL GET Request
** 21 - CRL POST Request
** ACL MANAGER
** 30 - ACL id create Request
** 31 - ACL id retrieve by device Request
** 32 - ACL id delete Request
** ACL INDIVIDUAL
** 40 - ACL individual get info Request
** 41 - ACL individual update ACE Request
** 42 - ACL individual delete Request
** ACL POLICY ENFORCEMENT
** 70 - ACL Check Permissions Request
** EXIT
** 99 - Exit cloud client

************************************************************
>> Enter Menu number:


혹 시 Cloud 예제 돌리 실때 서버와 클라이언트 간의 통신 절차를 간략하게 알려주실수 있으신가요?? 
부탁드려요

**Comment by Einstrasse** [comment]
IoTivity Cloud는 안드로이드 플랫폼에서는 지원안하고, 안드플랫폼은 클라이언트 역할만합니다. 리눅스로 빌드해야지 메시지 큐 퍼블리셔랑 서브스크라이버(mq_publisher, mq_subscriber)같은녀석들이 나타나고요.
클라우드라는 녀석이 리소스 디렉토리 역할을 포함하는데, 리소스 디렉토리라는 개념 자체가 나타난 것이 컴퓨팅과 네트워크 파워가 약한 lightweight 디바이스의 역할을 컴퓨팅과 네트워크 파워가 강한 rich device가 도와주는 개념이다 보니, 상대적으로 light device에서만 사용되는 안드로이드 플랫폼에서는 cloud기능을 제공하지 않는거죠.

저같은 경우에는 IoTivity 서버를 라즈베리파이, IoTivity Cloud를 랩탑으로 구성해서 홈 네트워크에 연결했고, 랩탑만 public ip 할당되게 포트포워딩을 했습니다.
이렇게 구성하면 외부 네트워크에서 IoTivity Client를 이용해서 홈에 있는 라즈베리파이의 IoTivity 서버에 접근가능하도록 구성했습니다. 인증은 github 계정을 통해서 했고 Iotivity Cloud에서 지원하는 3rd party oauth 인증을 이용해서 간단하게 구성해봤습니다.

그외에도 Cloud에서 제공하는 기능이 다양한걸로 알고 있는데 100% 활용해보지는 않아서 그 외적인 부분은 잘 모르겠고, Iotivity wiki에서 설명한 기본 예제만 적용해봤습니다. 그리고 Iotivity Cloud client는 안타깝게도 안드로이드 앱 형태로 된 예제는 제공하지 않는것 같더군요.

Iotivity Cloud를 이용해서 원격에서 홈 네트웤의 전등 제어하는 데모입니다.
youtu {dot} be {slash} _1IwBj0Dg8c

{dot}을 .로, {slash}를 /로 치환해서 사용하시면 됩니다.
